core/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -0,0 +1,263 @@
+final class StoreRecovery {
+
+    private final ESLogger logger;
+    private final ShardId shardId;
+
+    StoreRecovery(ShardId shardId, ESLogger logger) {
+        this.logger = logger;
+        this.shardId = shardId;
+    }
+
+    boolean recoverFromStore(final IndexShard indexShard, final boolean indexShouldExists, DiscoveryNode localNode) {
+        if (canRecover(indexShard)) {
+            if (indexShard.routingEntry().restoreSource() != null) {
+                throw new IllegalStateException("can't recover - restore source is not null");
+            }
+            try {
+                indexShard.recovering("from store", RecoveryState.Type.STORE, localNode);
+            } catch (IllegalIndexShardStateException e) {
+                // that's fine, since we might be called concurrently, just ignore this, we are already recovering
+                return false;
+            }
+            return recover(indexShard, () -> {
+                logger.debug("starting recovery from shard_store ...");
+                RecoveryState recoveryState = indexShard.recoveryState();
+                runRecoverFromStore(indexShard, indexShouldExists, recoveryState);
+                return recoveryState;
+            });
+        }
+        return false;
+    }
+
+    boolean restoreFromRepository(final IndexShard indexShard, IndexShardRepository repository) {
+        if (canRecover(indexShard)) {
+            if (indexShard.routingEntry().restoreSource() == null) {
+                throw new IllegalStateException("can't restore - restore source is null");
+            }
+            try {
+                indexShard.recovering("from snapshot", RecoveryState.Type.SNAPSHOT, indexShard.routingEntry().restoreSource());
+            } catch (IllegalIndexShardStateException e) {
+                // that's fine, since we might be called concurrently, just ignore this, we are already recovering
+                return false;
+            }
+            return recover(indexShard, () -> {
+                logger.debug("restoring from {} ...", indexShard.routingEntry().restoreSource());
+                RecoveryState recoveryState = indexShard.recoveryState();
+                restore(indexShard, recoveryState, repository);
+                return recoveryState;
+            });
+        }
+        return false;
+
+    }
+
+    private boolean canRecover(IndexShard indexShard) {
+        if (indexShard.state() == IndexShardState.CLOSED) {
+            // got closed on us, just ignore this recovery
+            return false;
+        }
+        if (!indexShard.routingEntry().primary()) {
+            throw new IndexShardRecoveryException(shardId, "Trying to recover when the shard is in backup state", null);
+        }
+        return true;
+    }
+
+    /**
+     * Recovers the state of the shard from the gateway.
+     */
+    private boolean recover(final IndexShard indexShard, Supplier<RecoveryState> recoveryRunnable) throws IndexShardRecoveryException {
```

Comment:
Why do we pass in the `indexShard` and a supplier of the recoverystate instead of getting it from the `indexShard` inside of the method? Just for easier mocking?